### PR TITLE
Ability to change colour theme.

### PR DESCRIPTION
**Pull request by muchweb**
~~Work in progress~~ :white_check_mark: Stable enough to merge

---

 - [x] Add two new schemes: 'all light', 'all dark' and accessibility 'high contrast'
 - [x] Tidy up loads of colour variables, that were spread across code
 - [x] Ability to choose from `default`, `dark`, `light` and `highcontrast` themes form command line by using `--theme` option
 - [x] Button styles
 - [x] Separate themes from main file
 - [x] Update usage manpage
 - [x] Ability to change theme from settings
 - [x] WIN32 \**duh*\* — ~~**HELP NEEDED** :boom:~~ — thanks to @GrayHatter
 - [x] Add status circles theming support
 - [x] Ability to theme buttons
 - ~~Ability to theme scrollbars~~ I think that it is OK to leave this for later as it would require a rewrite

---

### `utox`  or `utox --theme default`

utox-chat8-default

### `utox --theme light`

utox-chat7-light

### `utox --theme dark`

utox-chat7-dark

### `utox --theme highcontrast`

utox-chat7-highcontrast

### CODE REVIEW

**Comment by muchweb** [comment]
This will close https://github.com/notsecure/uTox/issues/796

**Comment by muchweb** [comment]
@GrayHatter, would you be able to help with this on android/win32?

File: android/gl.c [hunk 1]
```diff
@@ -30,7 +30,7 @@ fragment_shader[] =
         "gl_FragColor = (texture2D(samp, x) + vec4(k2, 0.0)) * vec4(k, 1.0);"
     "}";
 
-static GLuint prog, white;
+static GLuint prog, COLOUR_MAIN_BACKGROUND;
```
Comment (GrayHatter) [inline android/gl.c]: Can you use color instead of colour?

File: theme.c [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    switch (loadtheme) {
+    case THEME_DARK:
+        // Chat area colours
+        COLOUR_MAIN_BACKGROUND = 0x333333;
+        COLOUR_MAIN_FOREGROUND = 0xdfdfdf;
+        COLOUR_MAIN_FOREGROUND_SECONDARY = 0xbbbbbb;
+        COLOUR_MAIN_ACTIONTEXT = 0x27a9bc;
+        COLOUR_MAIN_URLTEXT = COLOUR_MAIN_ACTIONTEXT;
+        COLOUR_MAIN_QUOTETEXT = 0x55b317;
+        // Buddy list colours
```
Comment (GrayHatter) [inline theme.c]: s/buddy/friend

File: ui.c [hunk 3]
```diff
@@ -65,12 +65,11 @@ uint32_t status_color[] = {
...
 static void drawself(void)
 {
-    //40x40 self icon at 10,10
-    setcolor(button_name.mouseover ? C_STATUS : WHITE);
+    setcolor(!button_name.mouseover ? COLOUR_MENU_FOREGROUND : COLOUR_MENU_ACTIVE_FOREGROUND);
     setfont(FONT_SELF_NAME);
     drawtextrange(SELF_NAME_X, SELF_STATUS_X, SELF_NAME_Y, self.name, self.name_length);
 
-    setcolor(button_statusmsg.mouseover ? C_GRAY2 : C_STATUS);
+    setcolor(!button_statusmsg.mouseover ? COLOUR_MENU_FOREGROUND : COLOUR_MENU_ACTIVE_FOREGROUND);
```
Comment (GrayHatter) [inline ui.c]: why did you choose to !inverse this?

File: ui.c [hunk 4]
```diff
@@ -301,21 +301,21 @@ static void drawsettings_content(int UNUSED(x), int y, int UNUSED(w), int UNUSED
...
 static void background_draw(PANEL *UNUSED(p), int UNUSED(x), int UNUSED(y), int width, int height)
 {
-    drawrect(0, 0, LIST_RIGHT, LIST_Y - 1, LIST_DARK);
-    drawhline(0, LIST_Y - 1, LIST_RIGHT, LIST_EDGE);
-    drawrect(0, LIST_Y, LIST_RIGHT, height + LIST_BOTTOM, LIST_MAIN);
-    drawrect(0, height + LIST_BOTTOM, LIST_RIGHT, height, LIST_DARK);
-
+    // Current user avatar & name background
+    drawrect(0, 0, LIST_RIGHT, LIST_Y, COLOUR_MENU_BACKGROUND);
+    // Friend list ('roaster') background
```
Comment (GrayHatter) [inline ui.c]: roaster?

File: ui.c [hunk 5]
```diff
@@ -352,7 +352,7 @@ SCROLLABLE scroll_list = {
     .panel = {
         .type = PANEL_SCROLLABLE,
     },
-    .color = LIST_DARK,
+    .color = C_SCROLL,
```
Comment (GrayHatter) [inline ui.c]: is this the color you wanted to use?

File: utox.1 [hunk 6]
```diff
@@ -3,7 +3,8 @@
 µTox \- Lightweight Tox client
 
 .SH SYNOPSIS
-utox [OPTIONS]
+uasge: utox [--portable] [--theme <default|dark|light>]
```
Comment (GrayHatter) [inline utox.1]: maybe remove default and add high contrast?

File: xlib/main.c [hunk 7]
```diff
@@ -884,19 +882,58 @@ static int systemlang(void)
...
-        if(!strcmp(argv[1], "--version")) {
-            debug("%s\n", VERSION);
-            return 0;
-        } else if(!strcmp(argv[1], "--portable")) {
-            debug("Launching uTox in portable mode: All data will be saved to the tox folder in the current working directory\n");
-            utox_portable = 1;
-        } else {
-            debug("Valid arguments are: --version and --portable (launches uTox in portable mode)\n");
-            return 0;
+    parse_args_wait_for_theme = 0;
```
Comment (GrayHatter) [inline xlib/main.c]: can you rewrite this?

remove parse_wait_for.... completely, bury it into the strcmp --theme then check argv[(i+1)) for your options with a case(){}, then if you get data you don't expect, throw an error and set default

**Comment by GrayHatter** [comment]
all the colors in theme.c need a `// human readable color name`

**Comment by GrayHatter** [comment]
@muchweb pull this commit in to fix windows, then take a look at the setting page, some text is missing.

You can also see what I mean about moving argvc names into the --theme section

Additionally, why did you change the default highlighting from light blue to red?

File: ui_dropdown.c [hunk 8]
```diff
@@ -201,6 +213,13 @@ dropdown_logging = {
     .userdata = noyesdrops
 },
 
+dropdown_theme = {
```
Comment (GrayHatter) [inline ui_dropdown.c]: needs to update with argc/v calls

File: xlib/main.c [hunk 9]
```diff
@@ -884,19 +882,58 @@ static int systemlang(void)
     return ui_guess_lang_by_posix_locale(str, DEFAULT_LANG);
 }
 
+_Bool parse_args_wait_for_theme;
+char theme;
```
Comment (GrayHatter) [inline xlib/main.c]: I moved this to theme.h, but only in win32/*

File: xlib/main.c (hunk 7 above)
Comment (muchweb) [inline xlib/main.c]: `argv[i+1]` can easily lead to an overflow if this is the last argument. If I am going to remove the 'parse_args_wait_for_theme', will need to add extra check for array length.

File: android/gl.c (hunk 1 above)
Comment (muchweb) [inline android/gl.c]: Updated.

File: theme.c (hunk 2 above)
Comment (muchweb) [inline theme.c]: Updated

File: ui.c (hunk 3 above)
Comment (muchweb) [inline ui.c]: It made sense to first specify normal text colour, and active state after.

File: utox.1 (hunk 6 above)
Comment (muchweb) [inline utox.1]: Updated

**Comment by muchweb** [comment]
> all the colors in theme.c need a `// human readable color name`

I have updated variable names, now it should be more straightforward.

File: ui.c (hunk 4 above)
Comment (muchweb) [inline ui.c]: I am laughing right now! :joy_cat:

File: xlib/main.c (hunk 9 above)
Comment (muchweb) [inline xlib/main.c]: Updated. I've removed this from xlib `main.c` too

File: ui_dropdown.c (hunk 8 above)
Comment (muchweb) [inline ui_dropdown.c]: Resolved.

File: ui.c (hunk 5 above)
Comment (muchweb) [inline ui.c]: I've decided not to do scroll bar styling as part of this PR as it's likely to require a rewrite

**Comment by muchweb** [comment]
@GrayHatter, @irungentoo, regarding colour issues in wi32, could this be wrong data type? I have updated variables to use `uint32_t`.

Also, in the dark theme, for me it is not showing red highlight, but blue, as you can see on the screenshots. Were you testing this on win32?

**Comment by muchweb** [comment]
This is original `RGB` macro for win32, defined in `WinGDI.h`
```c
#define RGB(r,g,b)      ((COLORREF)(((BYTE)(r)|((WORD)((BYTE)(g))<<8))|(((DWORD)(BYTE)(b))<<16)))
```
It seems to do exactly the same thing.

**Comment by GrayHatter** [comment]
@muchweb I've added some commits back to my repo, can you pull those in and check to see if that fixes those problems?

It's also in PR #872

**Comment by muchweb** [comment]
@GrayHatter, I have pulled the `avoid buffer overflow or segfault issues`, but not the colour changes ones. Could you please check if my last type update fixes the issue (on this branch).

**Pull request merged**